### PR TITLE
[DOC] update multiclass example

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Relates to #6265

Use of AI tools:
- [ ] LLM tools were used to generate at least part of the content of this pull-request.
- [x] No LLM tools were used to generate any part of the content of this pull-request.

Changes proposed in this pull request:

- update multiclass example

### CODE REVIEW

File: nilearn/plotting/matrix/matrix_plotting.py [hunk 1]
```diff
@@ -223,9 +223,6 @@ def plot_matrix(
         Accepted linkage options for the clustering are 'single',
         'complete', and 'average'. True defaults to average linkage.
 
-        .. note::
-            This option is only available with SciPy >= 1.0.0.
-
```
Comment (Remi-Gau) [inline nilearn/plotting/matrix/matrix_plotting.py:267]: our minimal version is already > 1.0.0

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 2]
```diff
@@ -2,77 +2,84 @@
...
+
 from nilearn.maskers import NiftiMasker
 
-# For decoding, standardizing is often very important
 nifti_masker = NiftiMasker(
     mask_img=mask_filename,
     runs=run,
     smoothing_fwhm=4,
     memory="nilearn_cache",
+    standardize="zscore_sample",
```
Comment (Remi-Gau) [inline examples/02_decoding/plot_haxby_multiclass.py:64]: this used to be there in older version of the example and was implied by the comment above

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 3]
```diff
@@ -2,77 +2,84 @@
...
 
-# Remove the rest condition, it is not very interesting
-non_rest = y != "rest"
-y = y[non_rest]
-
-# Get the labels of the numerical conditions represented by the vector y
-unique_conditions, order = np.unique(y, return_index=True)
-# Sort the conditions by the order of appearance
-unique_conditions = unique_conditions[np.argsort(order)]
+print(pd.crosstab(y, run))
```
Comment (Remi-Gau) [inline examples/02_decoding/plot_haxby_multiclass.py]: I don't think we show in any example that the experimental design is balanced with same number of items per run... 

Also not mentioned in the dataset description 
https://nilearn.github.io/dev/modules/description/haxby2001.html

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 4]
```diff
@@ -94,51 +102,85 @@
...
 # Now we compute cross-validation scores
 # --------------------------------------
-from sklearn.model_selection import cross_val_score
+# The :term:`fMRI` data is acquired by runs,
+# and the noise is autocorrelated in a given run.
+# Hence, it is better to predict across runs when doing cross-validation.
+# To leave a run out, pass the cross-validator object
+# to the cv parameter of decoder.
 
-cv_scores_ovo = cross_val_score(svc_ovo, X, y, cv=5, verbose=1)
```
Comment (Remi-Gau) [inline examples/02_decoding/plot_haxby_multiclass.py:100]: is there a reason why the CV as not made relying on the runs?

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 5]
```diff
@@ -94,51 +102,85 @@
...
+show()
+
 # %%
-# Plot a confusion matrix
-# -----------------------
-# We fit on the first 10 runs and plot a confusion matrix on the
-# last 2 runs
+# Plot the confusion matrices
+# ---------------------------
+# We fit on the first 10 runs and plot a confusion matrix on the last 2 runs.
```
Comment (Remi-Gau) [inline examples/02_decoding/plot_haxby_multiclass.py]: BTW should we also in theory do a CV for the confusion matrix?

Comment (Remi-Gau) [reply examples/02_decoding/plot_haxby_multiclass.py]: actually implemented it

File: examples/02_decoding/plot_haxby_multiclass.py (hunk 4 above)
Comment (bthirion) [reply examples/02_decoding/plot_haxby_multiclass.py:100]: No good reason AFAICT

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 6]
```diff
@@ -90,55 +99,100 @@
...
+# and the noise is autocorrelated in a given run.
+# Hence, it is better to predict across runs when doing cross-validation.
+# To leave a run out, pass the cross-validator object
+# to the cv parameter of decoder.
 
-cv_scores_ovo = cross_val_score(svc_ovo, X, y, cv=5, verbose=1)
+from sklearn.model_selection import LeaveOneGroupOut, cross_val_score
 
-cv_scores_ova = cross_val_score(svc_ova, X, y, cv=5, verbose=1)
+cv = LeaveOneGroupOut()
```
Comment (bthirion) [inline examples/02_decoding/plot_haxby_multiclass.py]: IIRC there are 12 runs, which makes the cross-val loop longer.
This may have been the reason for CV=5.
Maybe take pairs of runs instead.

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 7]
```diff
@@ -90,55 +99,100 @@
...
+cv_scores_ovo = cross_val_score(svc_ovo, X, y, cv=cv, verbose=1, groups=run)
+cv_scores_ovo
+
+# %%
+cv_scores_ova = cross_val_score(svc_ova, X, y, cv=cv, verbose=1, groups=run)
+cv_scores_ova
+
+# %%
+print("OvO:", cv_scores_ovo.mean().round(decimals=3))
+print("OvA:", cv_scores_ova.mean().round(decimals=3))
```
Comment (bthirion) [inline examples/02_decoding/plot_haxby_multiclass.py:133]: The accuracy is now 15% lower...
Maybe the non-stationarity across sessions is the culprit. Using session pairs may improve this.

**Comment by Remi-Gau** [comment]
rendered example: https://output.circle-artifacts.com/output/job/27915ec0-f623-4306-b28a-1e3e9eccddf4/artifacts/0/dev/auto_examples/02_decoding/plot_haxby_multiclass.html#sphx-glr-auto-examples-02-decoding-plot-haxby-multiclass-py

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 8]
```diff
@@ -2,77 +2,84 @@
...
 mask_filename = haxby_dataset.mask
 
-# Load the behavioral data that we will predict
+print(f"Mask nifti images are located at: {mask_filename}")
+print(f"Functional nifti images are located at: {func_filename}")
+
+
+# %%
+# We load the behavioral data that we will predict and
+# remove the rest condition, as it is of no interest to us.
```
Comment (emdupre) [inline examples/02_decoding/plot_haxby_multiclass.py]: A bit confusing to comment here since the printed output immediately below includes rest. Could we add a new comment after the pd crosstab call to note that we are removing the rest condition as of no-interest / to look at balanced classes ?

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 9]
```diff
@@ -2,77 +2,84 @@
...
-unique_conditions, order = np.unique(y, return_index=True)
-# Sort the conditions by the order of appearance
-unique_conditions = unique_conditions[np.argsort(order)]
 
 # %%
 # Prepare the :term:`fMRI` data
 # -----------------------------
+# We extract the data with a NiftiMasker.
+# For decoding, standardizing is often very important,
+# so we set ``standardize="zscore_sample"``.
```
Comment (emdupre) [inline examples/02_decoding/plot_haxby_multiclass.py:56]: Should we explain what `zscore_sample` does ; i.e.  Timeseries are shifted to zero mean and scaled to unit variance. Uses sample std ?

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 10]
```diff
@@ -2,77 +2,84 @@
...
+# %%
 # Remove the "rest" condition
 X = X[non_rest]
 run = run[non_rest]
 
 # %%
 # Build the decoders, using scikit-learn
 # --------------------------------------
+# Nilearn does not have dedicate multiclass estimators,
+# so we rely on those from sklearn.
```
Comment (emdupre) [inline examples/02_decoding/plot_haxby_multiclass.py]:
```suggestion
# as we can directly use those from sklearn.
```

Comment (emdupre) [reply examples/02_decoding/plot_haxby_multiclass.py]: Light framing suggestion, since I think this is by design

File: examples/02_decoding/plot_haxby_multiclass.py (hunk 4 above)
Comment (emdupre) [reply examples/02_decoding/plot_haxby_multiclass.py:100]: I think this was nodded to with "We fit on the first 10 runs and plot a confusion matrix on the last 2 runs" but the example didn't really discuss CV much beforehand. I think it's a good addition !

File: examples/02_decoding/plot_haxby_multiclass.py (hunk 9 above)
Comment (Remi-Gau) [reply examples/02_decoding/plot_haxby_multiclass.py:56]: I think we should try to make sure it's mentioned every time we use it explicitly.
However... It will become the new default in a lot of places (especially for the decoding estimators) in 0.15.0 (so VERY soon)

File: examples/02_decoding/plot_haxby_multiclass.py (hunk 8 above)
Comment (Remi-Gau) [reply examples/02_decoding/plot_haxby_multiclass.py]: sounds good

note: in general I find that many of our examples spend too little time describing the data they use

will try to find a way to provide a succint and reusable description of the structure of haxby dataset

Comment (emdupre) [reply examples/02_decoding/plot_haxby_multiclass.py]: Would you add it at 
> :ref:`dataset description <haxby_dataset>` ?

Would be good to add that ref here !

**Review by bthirion — COMMENTED** [review]
The PR LGTM but we still have an issue with documentation generation...

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 11]
```diff
@@ -2,78 +2,85 @@
 The haxby dataset: different multi-class strategies
```
Comment (hndgzkn) [inline examples/02_decoding/plot_haxby_multiclass.py]:
```suggestion
The Haxby dataset: different multi-class strategies
```

File: examples/02_decoding/plot_haxby_multiclass.py (hunk 8 above)
Comment (Remi-Gau) [reply examples/02_decoding/plot_haxby_multiclass.py]: removing the following for now: it's the best example where to have this

print(pd.crosstab(labels["labels"], labels["chunks"]))

File: examples/02_decoding/plot_haxby_multiclass.py (hunk 7 above)
Comment (Remi-Gau) [reply examples/02_decoding/plot_haxby_multiclass.py:133]: switching to using a 5 fold CV: though I am omitting the "we chose this CV scheme because it gives us the best results" as it feels a bit of p-hacky justification 😜

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 12]
```diff
@@ -2,52 +2,58 @@
...
 ===================================================
 
-We compare one vs all and one vs one multi-class strategies: the overall
-cross-validated accuracy and the confusion matrix.
+In this example,
+we compare one vs all and one vs one multi-class strategies:
+the overall cross-validated accuracy and the confusion matrix.
```
Comment (hndgzkn) [inline examples/02_decoding/plot_haxby_multiclass.py]: It might be good to add a link to [multi-class classification](https://scikit-learn.org/stable/modules/multiclass.html#multiclass-classification).

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 13]
```diff
@@ -2,52 +2,58 @@
...
 """
 
 import numpy as np
 import pandas as pd
 
 from nilearn import datasets
-from nilearn.plotting import plot_matrix, show
 
 # %%
 # Load the Haxby data dataset
```
Comment (hndgzkn) [inline examples/02_decoding/plot_haxby_multiclass.py]:
```suggestion
# Load the Haxby dataset
```

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 14]
```diff
@@ -2,52 +2,58 @@
...
 # ---------------------------
 
-
-# By default 2nd subject from haxby datasets will be fetched.
```
Comment (hndgzkn) [inline examples/02_decoding/plot_haxby_multiclass.py:21]: This comment seems useful to me.

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 15]
```diff
@@ -2,52 +2,58 @@
...
 mask_filename = haxby_dataset.mask
 
-# Load the behavioral data that we will predict
+print(f"Mask nifti images are located at: {mask_filename}")
+print(f"Functional nifti images are located at: {func_filename}")
+
+
+# %%
+# We load the behavioral data that we will predict and
+# remove the rest condition, as it is of no interest to us.
```
Comment (hndgzkn) [inline examples/02_decoding/plot_haxby_multiclass.py]:
```suggestion
# remove the ``rest`` condition, as it is of no interest to us.
```

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 16]
```diff
@@ -59,21 +65,25 @@
...
 
+# %%
 # Remove the "rest" condition
 X = X[non_rest]
 run = run[non_rest]
 
 # %%
 # Build the decoders, using scikit-learn
 # --------------------------------------
+# Nilearn does not have dedicate multiclass estimators,
```
Comment (hndgzkn) [inline examples/02_decoding/plot_haxby_multiclass.py]:
```suggestion
# Nilearn does not have dedicated multiclass estimators,
```

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 17]
```diff
@@ -2,52 +2,58 @@
 The haxby dataset: different multi-class strategies
 ===================================================
 
-We compare one vs all and one vs one multi-class strategies: the overall
-cross-validated accuracy and the confusion matrix.
+In this example,
+we compare one vs all and one vs one multi-class strategies:
```
Comment (hndgzkn) [inline examples/02_decoding/plot_haxby_multiclass.py]:
```suggestion
we compare ``one vs all`` and ``one vs one`` multi-class classification strategies:
```

File: examples/02_decoding/plot_haxby_multiclass.py (hunk 12 above)
Comment (hndgzkn) [reply examples/02_decoding/plot_haxby_multiclass.py]: Also below part can be rephrased with the above to be more clear:

the overall cross-validated accuracy and the confusion matrix.

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 18]
```diff
@@ -2,52 +2,58 @@
...
 func_filename = haxby_dataset.func[0]
 mask_filename = haxby_dataset.mask
 
-# Load the behavioral data that we will predict
+print(f"Mask nifti images are located at: {mask_filename}")
+print(f"Functional nifti images are located at: {func_filename}")
+
+
+# %%
+# We load the behavioral data that we will predict and
```
Comment (hndgzkn) [inline examples/02_decoding/plot_haxby_multiclass.py:34]: Actually data preparation starts here, no?

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 19]
```diff
@@ -59,21 +65,25 @@
...
 # Remove the "rest" condition
 X = X[non_rest]
 run = run[non_rest]
 
 # %%
 # Build the decoders, using scikit-learn
 # --------------------------------------
+# Nilearn does not have dedicate multiclass estimators,
+# as we can directly use those from sklearn.
 # Here we use a Support Vector Classification, with a linear kernel,
```
Comment (hndgzkn) [inline examples/02_decoding/plot_haxby_multiclass.py]:
```suggestion
# Here we use a Support Vector Classification (:class:`sklearn.svm.SVC`) from scikit-learn, with a linear kernel,
```

**Review by hndgzkn — APPROVED** [review]
LGTM with some suggestions.

**Review by bthirion — APPROVED** [review]
LGTM, thx

File: examples/02_decoding/plot_haxby_multiclass.py (hunk 14 above)
Comment (Remi-Gau) [reply examples/02_decoding/plot_haxby_multiclass.py:21]: OK let's keep it

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 20]
```diff
@@ -2,52 +2,58 @@
...
 from nilearn import datasets
-from nilearn.plotting import plot_matrix, show
 
 # %%
 # Load the Haxby data dataset
 # ---------------------------
 
-
-# By default 2nd subject from haxby datasets will be fetched.
 haxby_dataset = datasets.fetch_haxby()
```
Comment (Remi-Gau) [inline examples/02_decoding/plot_haxby_multiclass.py:24]:
```suggestion
# By default 2nd subject from haxby datasets will be fetched.
haxby_dataset = datasets.fetch_haxby()
```

File: examples/02_decoding/plot_haxby_multiclass.py (hunk 18 above)
Comment (Remi-Gau) [reply examples/02_decoding/plot_haxby_multiclass.py:34]: OK let's have single section with load and prepare the data

File: examples/01_plotting/plot_demo_more_plotting.py [hunk 21]
```diff
@@ -51,7 +51,7 @@
 
 from nilearn import datasets
 
-# haxby dataset to have anatomical image, EPI images and masks
+# The Haxby dataset to have anatomical image, EPI images and masks.
```
Comment (Remi-Gau) [inline examples/01_plotting/plot_demo_more_plotting.py:54]: made a quick check to make that haxby is capitalized in the codebase

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 22]
```diff
@@ -1,53 +1,61 @@
...
 
-# Load the behavioral data that we will predict
+print(f"Mask nifti images are located at: {mask_filename}")
+print(f"Functional nifti images are located at: {func_filename}")
+
+
+# %%
+# We load the behavioral data that we will predict and
+# remove the ``"rest"`` condition, as it is of no interest to us.
+# .. seealso::
```
Comment (Remi-Gau) [inline examples/02_decoding/plot_haxby_multiclass.py:37]:
```suggestion
#
# .. seealso::
```

File: examples/02_decoding/plot_haxby_multiclass.py [hunk 23]
```diff
@@ -91,55 +105,99 @@
...
 plt.title("Prediction: accuracy score")
 
+show()
+
+# %%
+# Plot the cross-validated confusion matrices
+# --------------------------------------------
+# Instead of fitting on a single train/test split, we use
+# :func:`sklearn.model_selection.cross_val_predict` with the same
+# leave-one-run-out cross-validation as above to get, for each sample,
```
Comment (Remi-Gau) [inline examples/02_decoding/plot_haxby_multiclass.py]:
```suggestion
# :func:`sklearn.model_selection.cross_val_predict` with the same
# cross-validation as above to get, for each sample,
```
